test/e2e/networking_perf.go

```diff
@@ -0,0 +1,249 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+// Tests network performance using iperf or other containers.
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/intstr"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// This is a magic number, don't rely on it as a well defined constant.
+	gceBandwidthBitsEstimate = int64(30000000000)
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = KubeDescribe("Networking IPerf [Experimental] [Feature:Network-Performance] ", func() {
+
+	f := NewDefaultFramework("network-perf")
+
+	// A few simple bandwidth tests which are capped by nodes.
+	runClientServerBandwidthMeasurement(f, 1, gceBandwidthBitsEstimate)
+	runClientServerBandwidthMeasurement(f, 5, gceBandwidthBitsEstimate)
+	runClientServerBandwidthMeasurement(f, 8, gceBandwidthBitsEstimate)
+
+	// As many clients as possible based on node count.
+	runClientServerBandwidthMeasurement(f, 10000, gceBandwidthBitsEstimate)
+})
+
+func runClientServerBandwidthMeasurement(f *Framework, numClient int, maxBandwidthBits int64) {
+
+	// TODO: Make this a function parameter, once we distribute iperf endpoints.
+	numServer := 1
+
+	It(fmt.Sprintf("should transfer ~ 1GB onto the service endpoint %v servers (maximum of %v clients)", numServer, numClient), func() {
+		nodes := ListSchedulableNodesOrDie(f.Client)
+		totalPods := len(nodes.Items)
+		// for a single service, we expect to divide bandwidth between the network.  Very crude estimate.
+		expectedBandwidth := int(float64(maxBandwidthBits) / float64(totalPods))
+		Expect(totalPods).NotTo(Equal(0))
+		appName := "iperf-e2e"
+		CreateServiceWithPods(f,
+			8001,
+			8002,
+			appName,
+			api.Container{
+				Name:  "iperf-server",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-s", "-p", "8001"},
+				Ports: []api.ContainerPort{{ContainerPort: 8001}},
+			},
+			// this will be used to generate the -service name which all iperf clients point at.
+			numServer, // 1 server
+		)
+		CreateServiceWithPods(f,
+			-1,
+			-1,
+			"iperf-e2e-cli", // this will be used to generate the -service name which all iperf clients point at.
+			api.Container{
+				Name:  "iperf-client",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-c", "service-for-" + appName, "-p", "8002", "--reportstyle", "C",
```

Comment:
please include a comment with the exact iperf command being run including the svc ip/name, for clarity.